### DISCUSSION TITLE
Phoenix and Webhooks mode.

### DISCUSSION THREAD

**Discussion by prtngn** [visciang/telegram#133] [General]
Hi. How I can use bot in Webhook mode with phoenix project?

**Comment by tmaabm** [comment]
I too wanted to ask a similar question. In my case, I was able to use the stateless mode in a Phoenix project but I'm also interested in the statefull mode.

My use case is to collect information sequentially and hence I need to keep state somewhere. More, the Telegram API doesn't allow to attach this state in `ForceReply` for example?

It's a great library BTW :tada:

**Reply by visciang** [reply]
Hi @tbm206.
I think your use case in not directly related to Phoenix / webhook interactions.

If you need a statefull chat, no matter if you interface via polling or webhook with the telegram servers, you can use a [Telegram.ChatBot](https://visciang.github.io/telegram/Telegram.ChatBot.html) behaviour. This will give you a "per chat" state out of the box.

Or even using the "stateless" Telegram.Bot behaviour, you can manage "your state" leveraging whatever persistency (db, ets/dets, etc.).

But keep in mind is that the Telegram.ChatBot behaviour it's more natural (and correct) way of interacting with a user over a 1-1 chat, a session, where every user interact with a single Telegram.ChatBot process.

**Reply by visciang** [reply]
Please open a new discussion if you need help on the statefull topic

**Comment by visciang** [comment]
Hi @prtngn.

The current webhook library interface is designed to start an HTTP server and setup a webhook 

The Telegram.Webhook operating mode starts an HTTP server (cowboy) that will receive the webook notification requests from the telegram server. The app is meant to be served over HTTP, a reverse proxy should be plance in front of it, facing the public network (public IP address) over HTTPS. Refer to the [Telegram.Webhook](https://visciang.github.io/telegram/Telegram.Webhook.html) module docs for more info.

The library takes care to setup Plug.Cowboy with a Telegram.Webhook.Router.

Personally I've not planned an integration with phoenix to serve the webhook endpoint there.

**Reply by prtngn** [reply]
Well, if I have time, I'll try to make a module to work with the phoenix.

**Reply by mxgrn** [reply]
I also would love to see this integrated with Phoenix, reusing its Plug pipelines.

**Reply by visciang** [reply]
Added in https://github.com/visciang/telegram/commit/9170d29afce4e9604ec42c867f9b2ce9b9880f8a some documentation about how to directly integrate the webhook endpoint in an existing phoenix / plug application facing the internet.

**Reply by visciang** [reply]
In version 2.0.0 the plug integration has been refined. Check out the documentation
